### ISSUE TITLE
Unresolved generic argument causes System.InvalidOperationException in GetValueNodeFromGenericArgument

### ISSUE THREAD

**Issue by Deadpikle** [mono/linker#1634]
## Environment

* .NET 5.0.100 AND 5.0.200-preview.20566.9 (tested with both; same error)
* macOS Catalina (10.15.7)

## Command

`dotnet publish "path/project.csproj" --configuration "Release" --framework "net5" --runtime "osx-x64" --output "dotnet-build/" --self-contained True /p:PublishTrimmed=true`

## Error

ILLink : error IL1012: IL Linker has encountered an unexpected error. Please report the issue at https://github.com/mono/linker/issues [/Users/myname/Documents/Projects/path/here/project.csproj]
  Fatal error in IL Linker
```
  Unhandled exception. System.InvalidOperationException: Operation is not valid due to the current state of the object.
     at Mono.Linker.Dataflow.ReflectionMethodBodyScanner.GetValueNodeFromGenericArgument(TypeReference genericArgument)
     at Mono.Linker.Dataflow.ReflectionMethodBodyScanner.ProcessGenericArgumentDataFlow(GenericParameter genericParameter, TypeReference genericArgument, IMemberDefinition source)
     at Mono.Linker.Steps.MarkStep.MarkGenericArgumentConstructors(IGenericInstance instance, IMemberDefinition sourceLocationMember)
     at Mono.Linker.Steps.MarkStep.MarkGenericArguments(IGenericInstance instance, IMemberDefinition sourceLocationMember)
     at Mono.Linker.Steps.MarkStep.GetOriginalType(TypeReference type, DependencyInfo reason, IMemberDefinition sourceLocationMember)
     at Mono.Linker.Steps.MarkStep.MarkType(TypeReference reference, DependencyInfo reason, IMemberDefinition sourceLocationMember)
     at Mono.Linker.Steps.MarkStep.MarkField(FieldDefinition field, DependencyInfo& reason)
     at Mono.Linker.Steps.MarkStep.MarkEntireTypeInternal(TypeDefinition type, Boolean includeBaseTypes, DependencyInfo& reason, IMemberDefinition sourceLocationMember, HashSet`1 typesAlreadyVisited)
     at Mono.Linker.Steps.MarkStep.MarkEntireAssembly(AssemblyDefinition assembly)
     at Mono.Linker.Steps.MarkStep.InitializeAssembly(AssemblyDefinition assembly)
     at Mono.Linker.Steps.MarkStep.Initialize()
     at Mono.Linker.Steps.MarkStep.Process(LinkContext context)
     at Mono.Linker.Pipeline.Process(LinkContext context)
     at Mono.Linker.Driver.Run(ILogger customLogger)
     at Mono.Linker.Driver.Main(String[] args)
```

## Notes

* Removing `/p:PublishTrimmed=True` fixes the issue.
* Would be happy to jump on a Discord or other channel to debug "live" with someone if that's helpful
* App runs just fine with `dotnet run`
* Possibly related: #1583, #1559 

## Further investigation

I followed the steps in [this comment](https://github.com/mono/linker/issues/1583#issuecomment-721735037) in 1583. It gave me this:

```
ILLink: error IL1010: Assembly 'Avalonia.Animation, Version=0.9.999.0, Culture=neutral, PublicKeyToken=null' cannot be loaded due to failure in processing 'Avalonia.Base, Version=0.9.999.0, Culture=neutral, PublicKeyToken=null' reference
Mono.Linker.LinkerFatalErrorException: ILLink: error IL1010: Assembly 'Avalonia.Animation, Version=0.9.999.0, Culture=neutral, PublicKeyToken=null' cannot be loaded due to failure in processing 'Avalonia.Base, Version=0.9.999.0, Culture=neutral, PublicKeyToken=null' reference
 ---> Mono.Linker.LinkerFatalErrorException: ILLink: error IL1010: Assembly 'Avalonia.Base, Version=0.9.999.0, Culture=neutral, PublicKeyToken=null' cannot be loaded due to failure in processing 'System.Reactive, Version=4.4.0.0, Culture=neutral, PublicKeyToken=94bc3704cddfc263' reference
 ---> Mono.Linker.LinkerFatalErrorException: ILLink: error IL1009: Assembly 'System.Reactive, Version=4.4.0.0, Culture=neutral, PublicKeyToken=94bc3704cddfc263' reference 'System.Runtime.InteropServices.WindowsRuntime, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' could not be resolved
 ---> Mono.Cecil.AssemblyResolutionException: Failed to resolve assembly: 'System.Runtime.InteropServices.WindowsRuntime, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
 ---> System.IO.FileNotFoundException: Unable to find 'System.Runtime.InteropServices.WindowsRuntime.dll' or 'System.Runtime.InteropServices.WindowsRuntime.exe' file
   at Mono.Linker.DirectoryAssemblyResolver.Resolve(AssemblyNameReference name, ReaderParameters parameters)
   at Mono.Linker.AssemblyResolver.Resolve(AssemblyNameReference name, ReaderParameters parameters)
   at Mono.Linker.LinkContext.Resolve(IMetadataScope scope)
   at Mono.Linker.LinkContext.ResolveReferences(AssemblyDefinition assembly)
   --- End of inner exception stack trace ---
   at Mono.Linker.LinkContext.ResolveReferences(AssemblyDefinition assembly)
   at Mono.Linker.Steps.LoadReferencesStep.ProcessReferences(AssemblyDefinition assembly)
   at Mono.Linker.Steps.LoadReferencesStep.ProcessReferences(AssemblyDefinition assembly)
   --- End of inner exception stack trace ---
   at Mono.Linker.Steps.LoadReferencesStep.ProcessReferences(AssemblyDefinition assembly)
   at Mono.Linker.Steps.LoadReferencesStep.ProcessReferences(AssemblyDefinition assembly)
   --- End of inner exception stack trace ---
   at Mono.Linker.Steps.LoadReferencesStep.ProcessReferences(AssemblyDefinition assembly)
   at Mono.Linker.Steps.LoadReferencesStep.ProcessAssembly(AssemblyDefinition assembly)
   at Mono.Linker.Steps.BaseStep.Process(LinkContext context)
   at Mono.Linker.Pipeline.ProcessStep(LinkContext context, IStep step)
   at Mono.Linker.Pipeline.Process(LinkContext context)
   at Mono.Linker.Driver.Run(ILogger customLogger)
```

I can confirm that `System.Runtime.InteropServices.WindowsRuntime` does not exist on disk. Not sure why a windows runtime dll is needed on macOS? Perhaps this is ultimately an issue with [Avalonia](https://github.com/AvaloniaUI/Avalonia) and I need to update my version there?

**Comment by iXab3r** [comment]
Encountered the same problem(error-code and stack-trace vise) with WPF application. 

Application is using .NET 5 WPF and a library Blue.MVVM.Converter which is kinda old and was compiled for .NET 4.5 without any dependencies. All it has are a bunch of implementations of IValueConverter which were not changed since then.
The app compiles and works without any problems, but if you try to add trimming - ILLinker crashes with exactly the same stack 
trace without any additional info on which assembly or part of an app is causing the problem.
Ended up with checking binary MSBuild log and running ILLink step by hand.
Got the following error, which shows that the problem is in missing mscorlib of version 4.0,0.0 that Blue.MVVM.Converter library is referencing. Tried removing Blue.MVVM.Converter library trimming by adding TrimmerRootAssembly, but it did not help. Still looking for a solution.

>D:\Work\EyeAuras\Sources\PoeShared.UI>"C:\Program Files\dotnet\dotnet.exe" "C:\Program Files\dotnet\sdk\5.0.100\Sdks\Microsoft.NET.ILLink.Tasks\tools\netcoreapp3.0\illink.dll" -a  "C:\Users\Xab3r\.nuget\packages\blue.mvvm.converter\1.5.0\lib\net45\Blue.MVVM.Converter.dll"
ILLink: error IL1009: Assembly 'Blue.MVVM.Converter, Version=1.5.0.27868, Culture=neutral, PublicKeyToken=1bb1db17d606db59' reference 'mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089' could not be resolved
Mono.Linker.LinkerFatalErrorException: ILLink: error IL1009: Assembly 'Blue.MVVM.Converter, Version=1.5.0.27868, Culture=neutral, PublicKeyToken=1bb1db17d606db59' reference 'mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089' could not be resolved
 ---> Mono.Cecil.AssemblyResolutionException: Failed to resolve assembly: 'mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
 ---> System.IO.FileNotFoundException: Unable to find 'mscorlib.dll' or 'mscorlib.exe' file
   at Mono.Linker.DirectoryAssemblyResolver.Resolve(AssemblyNameReference name, ReaderParameters parameters)
   at Mono.Linker.AssemblyResolver.Resolve(AssemblyNameReference name, ReaderParameters parameters)
   at Mono.Linker.LinkContext.Resolve(IMetadataScope scope)
   at Mono.Linker.LinkContext.ResolveReferences(AssemblyDefinition assembly)
   --- End of inner exception stack trace ---
   at Mono.Linker.LinkContext.ResolveReferences(AssemblyDefinition assembly)
   at Mono.Linker.Steps.LoadReferencesStep.ProcessReferences(AssemblyDefinition assembly)
   at Mono.Linker.Steps.LoadReferencesStep.ProcessAssembly(AssemblyDefinition assembly)
   at Mono.Linker.Steps.BaseStep.Process(LinkContext context)
   at Mono.Linker.Pipeline.ProcessStep(LinkContext context, IStep step)
   at Mono.Linker.Pipeline.Process(LinkContext context)
   at Mono.Linker.Driver.Run(ILogger customLogger

Environment:
- .NET 5.0.100
- Windows

**Comment by vitek-karas** [comment]
The `InvalidOperationException` is VERY likely the same problem as in https://github.com/mono/linker/issues/1559. It has been fixed in 6, we're considering backporting it to 5 for a future servicing release.

The failure about `System.Runtime.InteropServices.WindowsRuntime.dll` is basically expected. It's one of the reasons why `--skip-unresolved true` is the default for the linker right now. 

The failure about mscorlib is a bit weird. .NET Core (or .NET 5) carries `mscorlib.dll` with it (for backward compat reasons), so it should be able to find it. You could grab the msbuild binlog and look at the linker command line - it should contain a `--reference <path>/mscorlib.dll` somewhere in it, does it?

**Comment by iXab3r** [comment]
Yes it has the following cmd line parameter:
`-reference "C:\Users\Xab3r\.nuget\packages\microsoft.netcore.app.runtime.win-x64\5.0.0\runtimes\win-x64\lib\net5.0\mscorlib.dll"

Maybe the problem is that assembly Blue.MVVM.Converters(that references mscorlib 4 instead of 5) has a strong name ?

**Comment by vitek-karas** [comment]
Strong names should not matter - both CoreCLR and linker only really care about simple name and version. Also `mscorlib.dll` in .NET 5 has version `4.0.0` (again for back compat reasons). I don't know, would probably need a repro to figure it out.

**Comment by iXab3r** [comment]
Narrowed down the error to a one old package - CsQuery.

To reproduce the error clone this and run "dotnet publish" 
https://github.com/iXab3r/ILLinkError1012Repro

**Comment by Deadpikle** [comment]
> The `InvalidOperationException` is VERY likely the same problem as in #1559. It has been fixed in 6, we're considering backporting it to 5 for a future servicing release.

:+1: Great. Thank you. Please let me know if I can be of any help. 😄 I'd definitely support the idea of a backport, although I know there are always other logistics involved!

**Comment by vitek-karas** [comment]
Thanks a lot for the repro @iXab3r - there's something else here. I tried it with latest build of mono/linker and it still fails 😢 . Looking into it. Whatever fix this will be I will try to push them all into servicing - it's should be doable, tooling crashes are a "must fix" basically (unless it's some kind of esoteric case, which this doesn't seem to be the case).

That said it doesn't repro the `mscorlib` failure - if you are still interesting and can get a repro for that one as well - I would love to look into it.

**Comment by vitek-karas** [comment]
The problem comes from the fact that `CsQuery` is a .NET Framework package. It has a hard dependency on `System.Web.Extensions.dll`, but that doesn't exist in .NET Core (or .NET 5). So even if linker could process it, at runtime if the related functionality would be requested it would not work anyway. Linker still must not crash on this though/

Linker has a bug where we don't handle unresolved type when seen in a generic argument (and possibly elsewhere). #1559 fixes a similar problem with array type being in this place, but it still left around this problem if the type is not resolvable at all. I'll use this issue to track this problem and fix it.

As for a workaround.. there's a `CsQuery.NETStandard` which doesn't have this problem. But I don't know if it's possible to use that in your app instead.

In general using .NET Framework assemblies on .NET Core/.NET 5 is "it might work, but no guarantees", this is one such case I guess.

**Comment by iXab3r** [comment]
@vitek-karas Thanks a lot for breaking down the issue ! 
You guys really did wonders on backward compatibility - the fact that such an old code still compiles and works on the latest version of framework is a huge success by itself, even without  latest compile-time features

**Comment by marek-safar** [comment]
The fix will be included in 5.0.2 service release
